File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cs
using System;

public class Solution
{
    int answer = 0;
    public int solution(int[] numbers, int target)
    {
        //- + 이진 트리
        //시작부터 끝까지 탐색
        //타겟 값인지 체크
        //맞으면 answer++;
        DFS(0, 0, numbers, target);
        return answer;
    }
    public void DFS(int depth, int sum, int[] numbers, int target)
    {
     
        if (depth == numbers.Length)
        {
            if (sum == target)
                answer++;
            return;
        }
        DFS(depth+1, sum+numbers[depth], numbers, target);
        DFS(depth+1, sum- numbers[depth], numbers, target);
    }

}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include<iostream>
using namespace std;

//숫자를 모두 써야함 = depth 체크 dfs
//+와 - 두가지 경우의 수
//depth 충족 시 sum == target 체크, 맞다면 answer++
    
void dfs(int depth, int sum, int target, int& answer , vector<int> numbers){
    if(depth == numbers.size()){
        if(sum == target)
            answer++;

        return ;
    }
        dfs(depth+1, sum + numbers[depth], target,answer, numbers);
        dfs(depth+1, sum - numbers[depth], target,answer, numbers);
}


int solution(vector<int> numbers, int target) {
    int answer = 0;
    dfs(0, 0, target, answer, numbers);
    return answer;
}

